core/message_pool.go

```diff
@@ -102,15 +133,20 @@ func (pool *MessagePool) Remove(c cid.Cid) {
 	pool.lk.Lock()
 	defer pool.lk.Unlock()
 
+	msg, ok := pool.pending[c]
+	if ok {
+		delete(pool.addressNonces, newAddressNonce(msg.message))
+	}
 	delete(pool.pending, c)
 	mpSize.Set(context.TODO(), int64(len(pool.pending)))
 }
 
 // NewMessagePool constructs a new MessagePool.
-func NewMessagePool(timer BlockTimer) *MessagePool {
+func NewMessagePool(api MessagePoolAPI) *MessagePool {
 	return &MessagePool{
-		timer:   timer,
-		pending: make(map[cid.Cid]*timedmessage),
+		api:           api,
+		pending:       make(map[cid.Cid]*timedmessage),
+		addressNonces: make(map[addressNonce]bool),
 	}
 }
 
```

Comment:
maybe remove the todo below, or turn it into issues if there is something specific still missing (i think not)